Allow for extension version to specified in friendly iframe embeds

alabiaga opened the pull request:
fixes 1/2 tasks in #12429

support 'extensions' on client side for friendlyIframeSpec, allowing for extensions and their versions.

related: #7217

alabiaga commented:
@lannka  @keithwrightbos @choumx friendly ping.

lannka commented on extensions/amp-a4a/0.1/test/test-amp-ad-utils.js:
```diff
@@ -18,6 +18,9 @@ import {data} from './testdata/valid_css_at_rules_amp.reserialized';
 import {getAmpAdMetadata} from '../amp-ad-utils';
 
 describe('getAmpAdMetadata', () => {
+  beforeEach(() => {
```
is this change needed?

alabiaga commented:
@lannka Thanks for the review. As far as your comment regarding the metadata for ads, customElementExtensions is definitely used, I saw documentation under amp-cache-modifications that it is deprecated but still in use and that "extensions" is supported..though i don't see anything in the code that supports and reads that data from that field. Perhaps the intention was for it to replace the cusomtElementExtensions, and the modifications have just not been made. @keithwrightbos Let me know if this is the case and whether this would be the preferred way to support versions with extensions and I can make that change...or perhaps I am wrong and didn't read the code correctly. Thanks

lannka commented:
@alabiaga that's what I see as well. Just want to confirm with @keithwrightbos because the server transformer needs to agree on the same format.
 
Meantime, please do have some manual test against `fake-ad.amp.html`, and play with extensions with different versions.

alabiaga commented:
@lannka awaiting Keith's input on this and I've pinged him. I will do the manual testing in the meantime. I've left the tests broken until I get his feedback. Many thanks for your feedback!

alabiaga commented:
confirmed that the extensions field in amp-ad-metadata is not used when determining which extensions to load but that the customElementExtensions is used. Thanks

keithwrightbos commented:
Apologies for the delayed response.  We are currently using customElementExtensions (see https://github.com/ampproject/amphtml/blob/master/extensions/amp-a4a/0.1/amp-a4a.js#L1601) which does not include the version, just the name of the extension.  However, we also return the extension in the format mentioned by @lannka and you should feel free to use it going forward.  Once fully deployed, we can remove sending the old customElementExtensions format.  Thanks

alabiaga commented:
spoke to @keithwrightbos, the client code was never modified to support this. the ad stack does support the extensions field for amp-ad-metadata when the ad is served. I will add this support on the client side.

jridgewell commented:
Oh, I think @newmuis actually did part of this?

alabiaga commented:
@jridgewell Correct, you cc'ed me in the message thread when @newmuis was asking about this bug. He did address this but in the context of reloading extensions. This PR is addressing the versioning bug in the context of FIE and ads. Thanks for mentioning.

alabiaga commented:
Friendly ping. this is ready for review.

jridgewell commented on src/service/extensions-impl.js:
```diff
@@ -445,13 +446,18 @@ export class Extensions {
-    extensionIds.forEach(extensionId => {
+    extensionData.forEach(extension => {
+      let extensionId;
+      if (extension && typeof extension == 'object') {
```
A4A should be passing down the `custom-element` fields directly. There's no reason for `extension-impl.js` to know anything about A4A objects.

jridgewell commented on src/service/extensions-impl.js:
```diff
@@ -631,6 +637,30 @@ export class Extensions {
+ */
+export function hasExtensionId(metadata, extensionId) {
+  let ext = [];
+  const {extensions, customElementExtensions} = metadata;
```
Again, this file should not have to be aware of A4A objects.

jridgewell commented on src/friendly-iframe-embed.js:
```diff
@@ -629,3 +641,28 @@ export function whenContentIniLoad(elementOrAmpDoc, hostWin, rect) {
+ * @param {!Window} win
+ */
+export function preloadExtensions(metaData, win) {
+  if (metaData.extensions && metaData.extensions.length) {
```
This "if one object type, do this, else if other type, do this" is icky. Why aren't we just standardizing a format to pass down?

alabiaga replied:
verified this is no longer used. Would've also helped if i read the comment here https://github.com/ampproject/amphtml/issues/7217#issuecomment-289822598, will remove customElementExtensions throughout codebase. Thanks

alabiaga commented:
@jridgewell @keithwrightbos ping. Added changes to gate on experiment. Also added a user warn log call to check that the old field isn't used.

related: https://github.com/ampproject/amphtml/pull/18738

jridgewell commented on extensions/amp-a4a/0.1/template-validator.js:
```diff
@@ -101,4 +94,52 @@ export class TemplateValidator extends Validator {
+      } else if (customElementExtensions && customElementExtensions.length) {
+        ext = customElementExtensions;
+      }
+      if (ext.indexOf(extensionId) == -1) {
```
I don't believe this can work in the `extensions` case, they're objects not strings.

jridgewell commented on extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js:
```diff
@@ -463,13 +463,21 @@ export class AmpAdNetworkAdsenseImpl extends AmpA4A {
+      } else if (customElementExtensions && customElementExtensions.length) {
+        ext = customElementExtensions;
+      }
+      if (ext.indexOf('amp-ad-exit') == -1) {
```
Ditto.

jridgewell commented on extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js:
```diff
@@ -913,13 +913,21 @@ export class AmpAdNetworkDoubleclickImpl extends AmpA4A {
+      } else if (customElementExtensions && customElementExtensions.length) {
+        ext = customElementExtensions;
+      }
+      if (ext.indexOf('amp-ad-exit') == -1) {
```
Ditto.

jridgewell commented on src/friendly-iframe-embed.js:
```diff
@@ -206,11 +220,24 @@ export function installFriendlyIframeEmbed(iframe, container, spec,
+      spec.customElementExtensions.forEach(extensionId => {
+        // src is not specified as legacy extensions will default to
+        // their 0.1 version. See extensions.installExtensionsInChildWindow.
+        ext.push({'custom-element': extensionId, 'src': null});
```
`version` instead of `src`?

jridgewell commented on src/friendly-iframe-embed.js:
```diff
@@ -629,3 +656,32 @@ export function whenContentIniLoad(elementOrAmpDoc, hostWin, rect) {
+        extensionDef => {
+          const extensionId = extensionDef['custom-element'];
+          const version = extensionDef['src'];
+          Services.extensionsFor(win)
```
Extract this part out of the loop (and if statements).

lannka commented:
@alabiaga are you still on it? or shall we close?

alabiaga commented:
@lannka unfortunately this CL has been stale as i haven't had time to fix the test. I'll try to resolve this week..otherwise i'll close it.

mrjoro commented:
@alabiaga is this PR still active?

alabiaga commented:
closing. these changes shouldn't be all too difficult to port over in another PR without the complications of merging.
